src/libexpr/primops.cc

```diff
@@ -1638,6 +1638,29 @@ static RegisterPrimOp primop_hashFile({
     .fun = prim_hashFile,
 });
 
+
+static void prim_readFileType(EvalState & state, const PosIdx pos, Value * * args, Value & v)
+{
+    auto path = realisePath(state, pos, *args[0]);
+    /* Enum DT_(DIR|LNK|REG|UNKOWN) */
+    auto type = getFileType(path);
+    v.mkString(
+        type == DT_REG ? "regular" :
+        type == DT_DIR ? "directory" :
+        type == DT_LNK ? "symlink" :
+        "unknown");
```

Comment:
This should be deduplicated with the same code in `readDir`. That's already done in the lazy-trees branch: https://github.com/NixOS/nix/pull/6530/files#diff-0f59bb6f197822ef9f19ceae9624989499d170c84dfdc1f486a8959bb4588cafR1654-R1662